src/modules/emote_menu/components/Emotes.jsx

```diff
@@ -0,0 +1,303 @@
+import React, {useCallback, useEffect, useMemo, useRef, useState} from 'react';
+import Icon from 'rsuite/lib/Icon/index.js';
+import classNames from 'classnames';
+import VirtualizedList from './VirtualizedList.jsx';
+import emoteStore from '../stores/index.js';
+import styles from '../styles/emotes.module.css';
+import Emote from './Emote.jsx';
+import Icons from './Icons.jsx';
+import {NavigationModeTypes, RowHeight, WindowHeight} from '../../../constants.js';
+
+const MAX_COLUMN_COUNT = 6;
+
+let navigationMode = NavigationModeTypes.MOUSE;
+
+function travelUp(rowColumnCounts, {x, y}, numBlocks = 1) {
+  let newY = Math.max(0, y - numBlocks);
+  let newYColumnCount = rowColumnCounts[newY];
+
+  // if we've reached a row without columns, we want to travel upwards again
+  if (newYColumnCount === 0) {
+    if (newY > 0) {
+      return travelUp(rowColumnCounts, {x, y}, numBlocks + 1);
+    }
+
+    // if we've reached the top, we have nowhere else to go.. the existing Y must be the first
+    newY = y;
+    newYColumnCount = rowColumnCounts[y];
+  }
+
+  return {
+    x: Math.min(newYColumnCount - 1, x),
+    y: newY,
+  };
+}
+
+function travelDown(rowColumnCounts, {x, y}, numBlocks = 1) {
```

Comment:
nit: pull this all out to a custom react hook. we can call it something like `useGridKeyboardNavigation`